keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -1023,26 +1024,125 @@ this enhancement.
 
 #### Enforcement Actions
 
-For phase 1, all violations implicitly result in a `deny` enforcement action.
+For parity with admission webhooks, a validating policy may also emit audit
+annotations and warnings:
 
-For phase 2, we intend to support multiple enforcement actions.
+- [Audit
+  annotations](https://github.com/kubernetes/kubernetes/blob/97bbf07d3f3f20332912ee411fdf75ce84425e28/staging/src/k8s.io/api/admission/v1/types.go#L142)
+  are key/value pairs included in the audit event for an admission request. The
+  audit annotation key supplied by the policy definition author will be prefixed
+  with the name of the `ValidatingAdmissionPolicy` and policy binding, e.g.:
+  `mypolicy.mygroup.example.com/mybinding.mygroup.example.com/<annotation-key>`.
+- [Warnings](https://kubernetes.io/blog/2020/09/03/warnings/#admission-webhooks)
+  are string messages that are returned to API clients. Warning are returned for
+  both requests that are accepted and requests that are rejected.
 
-Use cases:
+`ValidatingAdmissionPolicy` may declare audit annotations in the policy
+definition. E.g.:
 
-- Cluster admin would like to rollout a policies, sometimes in bulk, without
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+...
+spec:
+  ...
+  validations:
+    - expression: <expression>
+  auditAnnotations:
+    - includeWhen: <expression> # optional field
+      key: "my-audit-key"
+      valueExpression: <expression that evaluates to a string>
+```
+
+Additionally, `ValidatingAdmissionPolicyBinding` resource may control how
+admission is enforced. This is performed using a single field. E.g.:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicyBinding
+...
+spec:
+  enforcement: warn # optional field
+```
+
+- `deny`: Validation failures result in a denied request. (default beahvior if
+  field is unset)
+- `warn`: Validation failures are reported as warnings to the client.
```

Comment:
By putting the deny/warn choice here, from a policy perspective, there are only failures, and the policy author doesn't have visibility or control over whether they are fatal. For validating admission, maybe this makes sense... I'm less sure about mutating admission... mutating+allowing+warning seems more likely / reasonable, and is disjointed to express as `mutate+reject` in the policy and `enforce: warn` in the binding.

Reply:
>  putting the deny/warn choice here, from a policy perspective, there are only failures, and the policy author doesn't have visibility or control over whether they are fatal

This gets into the toughest part of this proposal.

The immediate goal is to offer safe rollout options, i.e. the ability "dryRun" a binding by first enabling it in a mode that is invisible to clients, then transitioning to a mode that produces client warnings, and then finally switching it to an enforcing mode.

But long term, if we want to offer policy authors control over whether a policy is fatal, then there could one field on the policy and another field on the binding that are involved in enforcement.

How about this option? Policy definitions eventually have `validations`, `auditAnnotation` and `warnings` stanzas (we don't have the `warnings`stanza yet).  Policy bindings have `validationEnforcement` and could in the future add similar controls for audit annotations and warnings.

I could update to match is this sounds reasonable.

Reply:
> I'm less sure about mutating admission... mutating+allowing+warning seems more likely / reasonable, and is disjointed to express as mutate+reject in the policy and enforce: warn in the binding.

I hadn't thought about mutating.  I'm not sure if we will want to even offer deny as an option for mutating. I'm not convinced it is particularly useful for webhooks either due to mutating webhook ordering.  But having warnings does seem more useful.  Maybe mutating will have stanzas like: `mutating`, `warnings`, `auditAnnotations` with enforcement controls for each stanza in the binding?  If/how we will offer rollout/dryrun for mutating is a problem I haven't thought about yet either..

Reply:
> I'm less sure about mutating admission...

I'm not sure we want to think of mutation and validation as parallel constructs:

* mutation is chained and recursed, validation is called in parallel and called once
* mutation operates on partial state (everything is subject to alteration by future mutations), validation sees the final state (because of this, users are discouraged from using mutation to inject invariants)
* mutation runs the risk of being non-idempotent in aggregate, validation does not

Also, the usage-pattern for mutations appears to mainly be admins adding defaulting behavior or writing logic to define annotations/labels that get auto-added, maybe extension authors adding defaults that openAPI schema doesn't support (though there are other ways to do this). Parameterization may be less valuable there.

TL;DR, thinking of validation and mutation as the same problem seems like a result of the history of them both being webhooks, but I'm not sure it's the best way forward. I expect mutation primitives may look different.

Reply:
> Policy definitions eventually have validations, auditAnnotation and warnings stanzas (we don't have the warningsstanza yet). 

I think we should preserve the ability to add a `warnings` stanza to a definition, to guard against being wrong, but we should wait for a concrete use case before actually adding it.

Keeping the intent in one place avoids the need to reason about how to combine different intents, and a binding author (who is making a policy concrete) is more likely to know the sensitivity of a policy.

Conversely, when might a definition author want to "warn" but not violate? If there is enough data to determine if the policy definition is violated, what's there to warn about? If there isn't, shouldn't the definition throw a violation to be safe?

I could see a definition author wanting to send various severity of violations (e.g. "syntax is wrong" vs. "this risks the integrity of your cluster", which a cluster admin might want to "warn" or "deny", respectively), but that could also be achieved by having two separate definitions.

Of course, all of this is glossing over mutation (which, per above, I consider a separate problem)

Reply:
Perhaps another direction is to make the `deny`, `warn`, `log` option be per-expression, and at the top level you could have a modifier to disable the whole thing, or send everything to logs for canarying, etc.

Reply:
A knob on the binding that overrides everything wholesale could work.

If we start dealing with "heirarchies" (e.g. the maximum action is `warn`, so change `deny` to `warn`, but `log` remains unchanged) things become brittle in that it implies there will always be a known ordering to actions taken.

IMO we could start off with only specifying action in the binding, and if it turns out people do need per-expression actions, it would be a backwards compatible change to add them in.

Reply:
I agree we should wait to see how things are used before we lock ourselves into a design by adding a `warnings` stanza. I've added future work notes about both per-expression enforcement and a `warnings` stanza, since we would have to decide which of those approaches to take.

Reply:
I think that's probably good, as long as we frame what is specified in the binding is a "maximum" (at *most*, we will deny or warn or log)

Reply:
I've added a note.